spec-tests/src/test/java/org/candlepin/spec/products/ProductVersioningSpecTest.java

```diff
@@ -0,0 +1,664 @@
+
+        // Another interesting case here. While it may appear that these should be different, the
+        // addition of the content to prod1 triggered the generation of a new product for org1 (prod3),
+        // leaving org2 as the only owner for prod2. That being the case, prod2 was updated in-place,
+        // allowing it to retain its UUID, and make this test look strange
+        // NOTE: The above will only be true when in-place updates are enabled. When they are disabled
+        // (as they are now), the UUIDs will not match, since we're always forking on any change.
+        assertNotEquals(prod4.getUuid(), prod1.getUuid()); // ^
+        assertNotEquals(prod4.getUuid(), prod2.getUuid()); // ^
+        assertNotEquals(prod4.getUuid(), prod3.getUuid());
+
+        // Actual test starts here
+        ContentDTO content3 = ownerContentApi.updateContent(owner2.getKey(), contentId,
+            new ContentDTO()
+            .name(contentName)
+            .id(contentId)
+            .label(contentLabel)
+            .type(contentType)
+            .vendor(contentVendor));
+        assertEquals(content3.getUuid(), content1.getUuid());
+        assertNotEquals(content3.getUuid(), content2.getUuid());
+
+        List<ProductDTO> prods = ownerProductApi.getProductsByOwner(owner1.getKey(), new ArrayList<>());
+        assertThat(prods).hasSize(1);
+        ProductDTO prod5 = prods.get(0);
+        assertNotEquals(prod5.getUuid(), prod1.getUuid());
+        assertNotEquals(prod5.getUuid(), prod2.getUuid());
+        assertEquals(prod5.getUuid(), prod3.getUuid());
+        assertNotEquals(prod5.getUuid(), prod4.getUuid());
+
+        prods = ownerProductApi.getProductsByOwner(owner2.getKey(), new ArrayList<>());
+        assertThat(prods).hasSize(1);
+        ProductDTO prod6 = prods.get(0);
+        assertNotEquals(prod6.getUuid(), prod1.getUuid());
+        assertNotEquals(prod6.getUuid(), prod2.getUuid());
+        assertEquals(prod6.getUuid(), prod3.getUuid());
+        assertNotEquals(prod6.getUuid(), prod4.getUuid());
+        assertEquals(prod6.getUuid(), prod5.getUuid());
+    }
+
+    @Test
+    public void shouldCleanupOrphansWithoutInterferingWithNormalActions() throws Exception {
+        // NOTE:
+        // This test takes advantage of the immutable nature of products with the in-place update branch
+        // disabled. If in-place updates are ever reenabled, we'll need a way to generate large numbers
+        // of orphaned products for this test.
+
+        OwnerDTO owner1 = ownerApi.createOwner(Owners.random());
+        OwnerDTO owner2 = ownerApi.createOwner(Owners.random());
+
+        String prefix = "test-product-";
+        int offset = 10000;
+        int length = 100;
+
+        // Repeat this test a few(ish) times to hopefully catch any synchronization error
+        for (int i = 0; i < 10; i++) {
+            final int base = offset;
+            List<String> o1Uuids = new ArrayList<>();
```

Comment:
Might want to use synchronized lists here to avoid failures as a result of local race conditions.